File: ThunderAI/ai/response_validator.py
import json
import re


Json = dict | list | str | int | float | bool | None

def _find_json(s: str) -> Json | None:
    candidates = re.findall(r'(\{.*\}|\[.*\])', s, re.DOTALL)
    candidates.insert(0, s)
    for candidate in candidates:
        try:
            return json.loads(candidate)
        except json.JSONDecodeError:
            continue
    return None

def clean_text_response(s: str) -> str:
    return (
        s.strip()
        .replace("**", "")
        .replace("*", "")
        .replace("#", "")
        .replace("`", "")
        .replace("~", "")
        .replace(">", "")
        .replace("---", "")
        .replace("___", "")
    )

def parsed_response(s: str) -> Json | str:
    json_data = _find_json(s)
    if json_data is None:
        return clean_text_response(s)
    return json_data

File: Supabase/supabase/functions/thunder/ai/response_validator.ts
export type Json = { [key: string]: Json } | Json[] | string | number | boolean | null;

function _find_json(s: string): Json | null {
  const candidates: string[] = s.match(/(\{.*\}|\[.*\])/gs) ?? [];
  candidates.unshift(s);

  for (const candidate of candidates) {
    try {
      return JSON.parse(candidate) as Json;
    } catch {
      // Continue trying next candidate.
    }
  }

  return null;
}

export function clean_text_response(s: string): string {
  return s
    .trim()
    .replaceAll("**", "")
    .replaceAll("*", "")
    .replaceAll("#", "")
    .replaceAll("`", "")
    .replaceAll("~", "")
    .replaceAll(">", "")
    .replaceAll("---", "")
    .replaceAll("___", "");
}

export function parsed_response(s: string): Json | string {
  const json_data = _find_json(s);
  if (json_data === null) {
    return clean_text_response(s);
  }

  return json_data;
}
